File: options_mm/engine/position.py
"""Position, P&L, and aggregate Greek tracking."""

from __future__ import annotations

from dataclasses import dataclass, field

from options_mm.engine.fill_sim import ContractKey, Trade
from options_mm.pricing.bsm import BSMResult


CONTRACT_MULTIPLIER = 100


@dataclass
class Position:
    quantity: int = 0
    cash: float = 0.0


@dataclass(frozen=True)
class RiskSnapshot:
    pnl: float
    net_delta: float
    net_gamma: float
    net_theta: float
    net_vega: float
    positions: dict[ContractKey, int]


@dataclass
class PositionManager:
    positions: dict[ContractKey, Position] = field(default_factory=dict)

    def record_trade(self, trade: Trade) -> None:
        position = self.positions.setdefault(trade.contract, Position())
        signed_qty = trade.signed_quantity
        position.quantity += signed_qty
        position.cash -= signed_qty * trade.price * CONTRACT_MULTIPLIER

    def mark_to_market(self, marks: dict[ContractKey, float]) -> float:
        pnl = 0.0
        for contract, position in self.positions.items():
            mark = marks.get(contract, 0.0)
            pnl += position.cash + position.quantity * mark * CONTRACT_MULTIPLIER
        return pnl

    def aggregate_greeks(self, greeks: dict[ContractKey, BSMResult]) -> dict[str, float]:
        totals = {"delta": 0.0, "gamma": 0.0, "theta": 0.0, "vega": 0.0}
        for contract, position in self.positions.items():
            result = greeks.get(contract)
            if result is None:
                continue
            multiplier = position.quantity * CONTRACT_MULTIPLIER
            totals["delta"] += result.delta * multiplier
            totals["gamma"] += result.gamma * multiplier
            totals["theta"] += result.theta * multiplier
            totals["vega"] += result.vega * multiplier
        return totals

    def snapshot(
        self,
        marks: dict[ContractKey, float],
        greeks: dict[ContractKey, BSMResult],
    ) -> RiskSnapshot:
        totals = self.aggregate_greeks(greeks)
        return RiskSnapshot(
            pnl=self.mark_to_market(marks),
            net_delta=totals["delta"],
            net_gamma=totals["gamma"],
            net_theta=totals["theta"],
            net_vega=totals["vega"],
            positions={
                contract: position.quantity
                for contract, position in self.positions.items()
                if position.quantity != 0
            },
        )

File: frontend/src/domain/position.ts
import {
  chainRowToContract,
  CONTRACT_MULTIPLIER,
  contractId,
  parseContractId,
  type ChainRow,
  type RiskSnapshot,
  type Trade,
} from './models';

export interface PositionState {
  quantity: number;
  cash: number;
}

export type PositionsMap = Record<string, PositionState>;

export interface GreekInputs {
  fair_value: number;
  delta: number;
  gamma: number;
  theta: number;
  vega: number;
}

/** Cash + mark-to-market: P&L = Σ(cash + qty × mark × multiplier). */
export function markToMarket(
  positions: PositionsMap,
  marks: Record<string, number>,
): number {
  let pnl = 0;
  for (const [id, position] of Object.entries(positions)) {
    const mark = position.quantity !== 0 ? (marks[id] ?? 0) : 0;
    pnl += position.cash + position.quantity * mark * CONTRACT_MULTIPLIER;
  }
  return pnl;
}

/**
 * Portfolio Greeks: Σ(qty × multiplier × greek) over open positions.
 * Delta = share-equivalent exposure; gamma/theta/vega in contract-dollar terms.
 */
export function aggregateGreeks(
  positions: PositionsMap,
  greeks: Record<string, GreekInputs>,
): Pick<RiskSnapshot, 'netDelta' | 'netGamma' | 'netTheta' | 'netVega'> {
  let netDelta = 0;
  let netGamma = 0;
  let netTheta = 0;
  let netVega = 0;

  for (const [id, position] of Object.entries(positions)) {
    if (position.quantity === 0) continue;
    const greek = greeks[id];
    if (!greek) continue;
    const multiplier = position.quantity * CONTRACT_MULTIPLIER;
    netDelta += greek.delta * multiplier;
    netGamma += greek.gamma * multiplier;
    netTheta += greek.theta * multiplier;
    netVega += greek.vega * multiplier;
  }

  return { netDelta, netGamma, netTheta, netVega };
}

export function snapshot(
  positions: PositionsMap,
  marks: Record<string, number>,
  greeks: Record<string, GreekInputs>,
): RiskSnapshot {
  const totals = aggregateGreeks(positions, greeks);
  const openPositions: Record<string, number> = {};

  for (const [id, position] of Object.entries(positions)) {
    if (position.quantity !== 0) {
      openPositions[id] = position.quantity;
    }
  }

  return {
    pnl: markToMarket(positions, marks),
    ...totals,
    positions: openPositions,
  };
}

export function recordTrade(positions: PositionsMap, trade: Trade): PositionsMap {
  const id = contractId(trade.contract);
  const current = positions[id] ?? { quantity: 0, cash: 0 };
  const signedQty = trade.direction === 'buy' ? trade.quantity : -trade.quantity;

  return {
    ...positions,
    [id]: {
      quantity: current.quantity + signedQty,
      cash: current.cash - signedQty * trade.price * CONTRACT_MULTIPLIER,
    },
  };
}

export function buildRiskInputs(
  ticker: string,
  chainRows: ChainRow[],
  positions: PositionsMap = {},
): { marks: Record<string, number>; greeks: Record<string, GreekInputs> } {
  const marks: Record<string, number> = {};
  const greeks: Record<string, GreekInputs> = {};

  for (const row of chainRows) {
    const contract = chainRowToContract(ticker, row);
    const id = contractId(contract);
    marks[id] = row.fair_value;
    greeks[id] = {
      fair_value: row.fair_value,
      delta: row.delta,
      gamma: row.gamma,
      theta: row.theta,
      vega: row.vega,
    };
  }

  // Ensure open positions use the latest chain marks even if strike keying differs slightly.
  for (const [id, position] of Object.entries(positions)) {
    if (position.quantity === 0 || marks[id] !== undefined) continue;
    const contract = parseContractId(id);
    const row = chainRows.find(
      (r) =>
        r.strike === contract.strike &&
        r.option_type === contract.option_type &&
        r.expiry === contract.expiry,
    );
    if (!row) continue;
    marks[id] = row.fair_value;
    greeks[id] = {
      fair_value: row.fair_value,
      delta: row.delta,
      gamma: row.gamma,
      theta: row.theta,
      vega: row.vega,
    };
  }

  return { marks, greeks };
}
